src/main/java/com/hierynomus/sshj/transport/cipher/GcmCipher.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C)2009 - SSHJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.sshj.transport.cipher;
+
+import net.schmizz.sshj.common.Buffer;
+import net.schmizz.sshj.common.SSHRuntimeException;
+import net.schmizz.sshj.transport.cipher.BaseCipher;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import java.security.GeneralSecurityException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+
+public class GcmCipher extends BaseCipher {
+
+    protected Mode mode;
+    protected boolean initialized;
+    protected CounterGCMParameterSpec parameters;
+    protected SecretKey secretKey;
+
+    public GcmCipher(int ivsize, int authSize, int bsize, String algorithm, String transformation) {
+        super(ivsize, authSize, bsize, algorithm, transformation);
+    }
+
+    protected Cipher getInitializedCipherInstance() throws GeneralSecurityException {
```

Comment:
Why is this necessary? If we construct a Cipher, it should also be initialized before calling out to `update` or `updateAAD`

Reply:
This is explainable with [RFC5647](https://tools.ietf.org/html/rfc5647) Section 7.1. Per RFC said, the IV will increase by 1 after each cipher invocation.

> With AES-GCM, the 12-octet IV is broken into two fields: a 4-octet fixed field and an 8-octet invocation counter field.  The invocation field is treated as a 64-bit integer and is incremented after each invocation of AES-GCM to process a binary packet.

And after the IV is changed it seems the cipher must be reinitialized with the updated GCMParameterSpec - simply increase the IV's invocation counter by 1 is not enough.

Reply:
Ok, agreed.